### ISSUE TITLE
Live streaming over WiFi

### ISSUE THREAD

**Issue by Gaederer** [google/oboe#387]
Hello, I am working on a project of live audio streaming over WiFi. 
In particular, I need to build two apps (a transmitter and a receiver), in order to let a person talk with the transmitter, so that a set of people can listen his voice using the receiver, being connected to the same WiFi network.
My current solution is experiencing a small delay, so I was considering to use Oboe on both apps to have better performances. My idea would be to create an AudioStream (with microphone as input source) on transmitter app, send it through network, then reproduce it on the receiver app through speakerphone or headset.
Is this technically feasible with Oboe? Can I convert AudioStream objects to/from byte arrays? Or should I look for a different solution?
Thanks for any advice.

**Comment by philburk** [comment]
You can use Oboe for this. You will need to convert the data from whatever
format you're using to one of the two formats Oboe can take, either 32-bit
floats or 16-bit integer audio data. The wireless part is up to you.

**Comment by Gaederer** [comment]
I am implementing the receiver app and I have a doubt. Given that I am getting audio data from wireless network through integer array, I am now wondering how to write this data to the "Playback AudioStream" I have created.
Reading the guide, I got that there are two ways to write data to an AudioStream:

- Using the write() method
- Filling the audioData buffer in AudioStreamCallback's onAudioReady() method

I read in the "Callback do's and don'ts" section that it should not be used for operations like streaming, so I guess I should go with write() method. Am I right?

**Comment by dturner** [comment]
Re-reading the "callback do's and don'ts" section I realise that the guidance isn't as clear as it should be (hopefully I've fixed that here: https://github.com/google/oboe/commit/0393ed439bcdad88f474274556665ca9cd19375f). To summarise: you shouldn't do anything which could block inside the callback. 

That said, your use case requires the lowest possible latency, and this requires using a callback. Here's what I'd do: 

- Read your data from the network on a separate thread and write it into a non-blocking queue (example here: https://github.com/google/oboe/blob/master/samples/RhythmGame/src/main/cpp/utils/LockFreeQueue.h)
- Inside `onAudioReady` read from the queue into the audio buffer

Note: You'll probably need to gracefully handle the case where there is no data in the queue (e.g. due to network bandwidth/connectivity issues) so that you don't get audible glitches.

**Comment by Gaederer** [comment]
Thanks for the clarification, your tips are really useful!

**Comment by dturner** [comment]
No problem. Feel free to reopen if you have more questions.

**Comment by Gaederer** [comment]
Hello, I need your help again.
I implemented my AudioEngine class, using a non-blocking queue and the onAudioReady() callback as you suggested.
I finally succeeded in reproducing the audio stream coming from Wi-Fi module, and I have to say that delay is really low: a very good result, compared to our previous implementation.
I have only a problem: the audio is like "disturbed", it seems not to be clean. People voice is a little distorted, I'd say "croaking". I think this is caused by the fact I'm not being able to handle the case of no data in buffer.
This is my callback at the moment:

```
oboe::DataCallbackResult AudioEngine::onAudioReady(oboe::AudioStream *audioStream, 
void *audioData, int32_t numFrames) {

    auto *outputData = static_cast<int16_t *>(audioData);
    int16_t nextFrame;

    // Read data from LockFreeQueue and write it to outputData
    for (int i = 0; i < numFrames; ++i) {
        bool popped = mAudioDataQueue.pop(nextFrame);

        if (popped) {
            outputData[i] = nextFrame;
        } else {
            // TODO Handle the case of no data
        }
    }

    return oboe::DataCallbackResult::Continue;
}
```


I'm not sure about what to do in the ELSE branch, that is the case of no data in the queue. I have tried putting zeros, but in that way I got a constant noise in background, like a beat.
I also was looking at how you use the memset() method in samples, but I didn't get a clear idea about the parameters I should pass. Is this the method I should use? Could you please clarify its usage?
Or is there something else I am missing?
Thanks in advance.

**Comment by dturner** [comment]
Please could you share your code for setting up the `AudioStream` using the `AudioStreamBuilder`?

**Comment by Gaederer** [comment]
Sure, here it is:

```
oboe::AudioStreamBuilder builder;
    builder.setCallback(this);
    builder.setDirection(oboe::Direction::Output);
    builder.setPerformanceMode(oboe::PerformanceMode::LowLatency);
    builder.setSharingMode(oboe::SharingMode::Exclusive);
    builder.setFormat(oboe::AudioFormat::I16);
    builder.setChannelCount(oboe::ChannelCount::Mono);
    builder.setContentType(oboe::ContentType::Speech);
    builder.setSampleRate(mSampleRate);

    oboe::Result result = builder.openStream(&mPlaybackStream);

    if (result != oboe::Result::OK) {
        LOGE("Error opening stream: %s", oboe::convertToText(result));
        // handle the error
    }
```

Please note that I defined mPlaybackStream in the header file (initializing it to nullptr) and that I set mSampleRate to the value passed by Wi-Fi module. I have seen that the value is hard-coded to 10000, I guess because this is the value set on the iOS transmitter app (I did not work on it). When I was not setting sample rate, I couldn't hear the transmitted audio.

**Comment by philburk** [comment]
>  bool popped = mAudioDataQueue.pop(nextFrame);

Should that be "(&nextFrame)" ?

> I have seen that the value is hard-coded to 10000, 

A sample rate of 10000 is pretty strange. It may not be well supported on your device.
Also you will not get a low latency stream when you specify an odd sample rate.

What Android version have you tested?
What devices are you running on?

You might want to not specify a sample rate, and then resample the audio to the actual device rate.

**Comment by Gaederer** [comment]
Hello, thanks for the response.

> Should that be "(&nextFrame)" ?

That variable is an int_16t, if I try with & I get a compiler error. I saw a similar usage in the RhythmGame sample. 

> What Android version have you tested?
What devices are you running on?

I am testing on an OnePlus6 running Android 9.

> A sample rate of 10000 is pretty strange. It may not be well supported on your device.
Also you will not get a low latency stream when you specify an odd sample rate.

I see that it is a strange value and I agree with you. I am wondering if it can be related to the audio glitches I am experiencing. I will try to modify that value on the iOS transmitter app.

But I have two more questions:

1) Is it also possible to resample the audio on the Android receiver app? Or the best option is changing the value on transmitter app?
2) I think I have definitely to handle the case of "no data" in the buffer (the ELSE branch in my code), but I really can't find a proper way to use the memset() method I see in the samples. Could you please explain the params I should pass to it?

**Comment by philburk** [comment]
> Should that be "(&nextFrame)" ?

Oh. I see that pop is defined as a template with (T &n), which is confusing. Not your fault.

Sending the data over WiFI at 10000 Hz is fine because that reduces the bandwidth.
For lowest latency, open the stream with unspecified sample rate then convert it to and from 10000 Hz.

memset needs the size in bytes = frameCount * samplesPerFrame * bytesPerSample
                   eg.   frameCount * 1 * sizeof(int16_t)

**Comment by Gaederer** [comment]
I'm stuck. This is my callback at the moment:

```
oboe::DataCallbackResult` AudioEngine::onAudioReady(oboe::AudioStream *audioStream, 
void *audioData, int32_t numFrames) {

    auto *outputData = static_cast<int16_t *>(audioData);
    int16_t nextFrame;
    int32_t framesRead = 0;

    // Read data from LockFreeQueue and write it to outputData
    for (int i = 0; i < numFrames; ++i) {
        bool popped = mAudioDataQueue.pop(nextFrame);

        if (popped) {
            outputData[i] = nextFrame;
            framesRead++;
        } else {
            // No data
             break;
        }
    }

    int32_t framesToPad = numFrames - framesRead;

    if (framesToPad > 0) {
        int32_t bytesPerFrame = audioStream->getChannelCount() * audioStream->getBytesPerSample();
        int16_t *padPos = outputData + framesRead * bytesPerFrame;
        auto size = static_cast<size_t>(framesToPad * audioStream->getChannelCount() * bytesPerFrame);

        // pad the buffer with zeros
        memset(padPos, 0, size);
    }

    return oboe::DataCallbackResult::Continue;
}
```

When I did not call memset(), I had the audio distortion (croaking voices).
Now that I am calling it, I have this continuous beat in background that is partially covering the audio.
Why this beat? What I am doing wrong in zero padding operation?
And, if I am not making mistakes in padding, where else should I focus to get a clean audio?

Thanks for your help.

**Comment by philburk** [comment]
> int16_t *padPos = outputData + framesRead * bytesPerFrame;

outputData is an int16_t* so you have to add the number of samples, **not** the number of bytes. (I often call channelCount "samplesPerFrame". Then the units are more clear.) Try:

   int32_t samplesRead = framesRead * audioStream->getChannelCount();
   int16_t *padPos = outputData + samplesRead;
    
> auto size = static_cast<size_t>(framesToPad * audioStream->getChannelCount() * bytesPerFrame);

A minor bug. bytesPerFrame already is times channelCount. So you have channelCount squared. Like one squared is one. Try:

    auto size = static_cast<size_t>(framesToPad * audioStream->getBytesPerFrame());

**Comment by dturner** [comment]
Closing due to inactivity. Feel free to reopen if you have more questions.

**Comment by Gaederer** [comment]
Hello again.
I completed the receiver app, thanks to your helpful advices. The performances are very good: low latency, almost no delay, clear audio.
Now I am working on the transmitter app. I started from the LiveEffect sample, since it is taking audio stream from input source and performing a playback effect.
I see the code inside the PlaybackStream's onAudioReady callback: data is read from RecordingStream and written to buffer. I guess I should *only* take this data and pass it to another layer in my app that sends it through an UDP socket.
But I was wondering about which is the best way to take and pass this data. Given that when data is read from RecordingStream, I push it to a LockFreeQueue, I have two ideas:

1) Read periodically from the Queue with a certain interval
2) Every time a bunch of data is read from RecordingStream, invoke a callback passing a short array populated by popping the Queue

Am I on the right path? Which would be the best approach here?

**Comment by philburk** [comment]
This was closed so I did not see your question until today. I will reopen this.

For VoIP you may not need to synchronize the output streams because they are not connected. 

Since you are writing the input data to a network stream, you might try this approach.
Open the input stream with no callback but with a large capacity. (Experiment with the size).
Launch a thread that reads the data from the input stream and writes it to the network stream. It could wake up at whatever the period is for your network packets.
This way you do not need a separate LockFreeQueue. Just use the stream buffer in AAudio.

**Comment by Gaederer** [comment]
Thanks for you suggestion.
I was doing something pretty similar, but I was still using a Queue.
Following your advice, I can listen the stream from the client side. The quality is pretty good (not perfect, though) and there is a very small delay. I guess I have to play with buffer size and wake up period.

**Comment by philburk** [comment]
Closing due to inactivity. Please open a new issue if needed.

**Comment by jerearaujo03** [comment]
Hello, I've followed all your comments, but I've not been able to hear the sound, only sound like white noise. I'm receiving int16_t data (stereo), so I'm receiving 2 bytes for sample for each channel.

I'm using this queue https://github.com/google/oboe/blob/master/samples/RhythmGame/src/main/cpp/utils/LockFreeQueue.h
Here is how I push to my queue
```
void AudioEngine::addToQueue(int16_t message) {
    audioQueue.push(message);
}
```

And here my callback:

```
DataCallbackResult
AudioEngine::onAudioReady(AudioStream *oboeStream, void *audioData, int32_t numFrames) {

    int16_t *outputData = static_cast<int16_t *>(audioData);
    int16_t nextFrame;
    int32_t framesRead = 0;

    for (int i = 0; i < numFrames; i++) {
        bool popped = audioQueue.pop(nextFrame);
        if (popped) {
            outputData[i] = nextFrame;
            framesRead++;
        } else {
            break;
        }
    }

    int32_t framesToPad = numFrames - framesRead;
    if (framesToPad > 0) {
        int32_t samplesRead = framesRead * stream->getChannelCount();
        int16_t *padPos = outputData + samplesRead;
        auto size = static_cast<size_t>(framesToPad * stream->getBytesPerFrame());

        // pad the buffer with zeros
        memset(padPos, 0, size);
    }
    return oboe::DataCallbackResult::Continue;
}
```

i've verified I'm pushing and popping the correct data from and to the queue.
Why it's not playing the sound correctly?

**Comment by philburk** [comment]
I suspect you may have an EndianNess issue.
Data sent over the network is usually BigEndian.
Android is usually LittleEndian. You may need to swap the bytes you get from the network.

**Comment by jerearaujo03** [comment]
Yes, I've tried with both, LittleEndian is the correct, because I can understand a bit the music I'm playing for testing, but it sounds horrible (also changing the tuning of the song)

**Comment by jerearaujo03** [comment]
I was doing a lot of debugging, and I realized the queue isn't fast enough. Is it posible?

**Comment by philburk** [comment]
I listened to the corrupted audio. I do not recognizes the distortion.
I notice in your callback that the for loop assumes the audio is mono because it processes one sample per frame. But the second part of the code allows for channels > 1. So there may be a channel issue.

Also I notice that when you push to the queue that it is not checking to see if the queue is full. So it may be dropping data and not throttling correctly.

**Comment by jerearaujo03** [comment]
I've tried sending the data from Java side as byte array, iterating the data on c++ side, and also tried sending the individual samples from Java side (one by one), and in c++ only adding to the queue, but both ways doesn't works. The sound is broken.

**Comment by jerearaujo03** [comment]
I'm sending stereo PCM 16 bits (signed). Also my queue is of type int16_t. Should it be int32_t for stereo, and I have to write first 2 bytes for L and last 2 bytes for R?

**Comment by jerearaujo03** [comment]
I've tried in mono, and it works fine, but the problem is when I try to play in stereo. What should I change in the callback for stereo playing?

**Comment by jerearaujo03** [comment]
Ok that was my problem, I solved doing this:
```
for (int i = 0; i < numFrames*stream->getChannelCount(); i++) {
    bool popped = audioQueue.pop(nextFrame);
    if (popped) {
        outputData[i] = nextFrame;
        framesRead++;
    } else {
        break;
    }
}
```
Now i'm multiplying `numFrames*stream->getChannelCount()`

**Comment by droid-dev-ganesh** [comment]
Hello, I am working on a project and I have the same requirement. I want to know how we can get the audio data back to an activity so that I can send this via socket to another android application. The data transfer should happen with minimum latency as I want to stream the data on other android app as quickly as possible. Please help me by sharing the logic or code snippet. Thank you in advance.

**Comment by dturner** [comment]
We're moving questions to StackOverflow. Please post there and use the `oboe` tag.

**Comment by Erik-Lehnsherr** [comment]
> Re-reading the "callback do's and don'ts" section I realise that the guidance isn't as clear as it should be (hopefully I've fixed that here: [0393ed4](https://github.com/google/oboe/commit/0393ed439bcdad88f474274556665ca9cd19375f)). To summarise: you shouldn't do anything which could block inside the callback.
> 
> That said, your use case requires the lowest possible latency, and this requires using a callback. Here's what I'd do:
> 
> * Read your data from the network on a separate thread and write it into a non-blocking queue (example here: https://github.com/google/oboe/blob/master/samples/RhythmGame/src/main/cpp/utils/LockFreeQueue.h)
> * Inside `onAudioReady` read from the queue into the audio buffer
> 
> Note: You'll probably need to gracefully handle the case where there is no data in the queue (e.g. due to network bandwidth/connectivity issues) so that you don't get audible glitches.

Dear Mr. Turner,

If possible, could you give some advices regarding the implementation of the followings from your previous reply?:
"Read your data from the network on a separate thread and write it into a non-blocking queue"

My current structure is to use DatagramPacket to transmit and receive the audio data. The data type of the buffer holding the incoming datagram is byte[]. It looks like I need to pass the data from the byte[] buffer through JNI (Java Native Interface) and then read the data into a non-blocking queue.

If I understand correctly, there is no byte[] data type in C++, so I have to convert the byte[] (jbyteArray in JNI) to the data type (ex: int16_t)?

However, I am not very sure how to implement such a conversion? Or even I am in a wrong direction?

Any suggestions or sample code would be highly appreciated!

Thank you so much!

Best!
